gym/dev_wrappers/torch_wrapper.py

```diff
@@ -0,0 +1,121 @@
+# This wrapper will convert torch inputs for the actions and observations to Jax arrays
+# for an underlying Jax environment then convert the return observations from Jax arrays
+# back to torch tensors.
+#
+# Functionality for converting between torch and jax types originally copied from
+# https://github.com/google/brax/blob/9d6b7ced2a13da0d074b5e9fbd3aad8311e26997/brax/io/torch.py
+# Under the Apache 2.0 license. Copyright is held by the authors
+
+"""Helper functions and wrapper class for converting between PyTorch and Jax."""
+import functools
+from collections import abc
+from typing import Any, Dict, Optional, Tuple, Union
+
+from jax._src import dlpack as jax_dlpack
+from jax.interpreters.xla import DeviceArray
+
+from gym import Env, Wrapper
+from gym.error import DependencyNotInstalled
+
+try:
+    import torch
+    from torch.utils import dlpack as torch_dlpack
+except ImportError:
+    raise DependencyNotInstalled("torch is not installed, run `pip install torch`")
+
+
+Device = Union[str, torch.device]
+
+
+@functools.singledispatch
+def torch_to_jax(value: Any) -> Any:
+    """Converts a PyTorch Tensor into a Jax DeviceArray."""
+    return value
+
+
+@torch_to_jax.register(torch.Tensor)
+def _torch_to_jax(value: torch.Tensor) -> DeviceArray:
+    """Converts a PyTorch Tensor into a Jax DeviceArray."""
+    tensor = torch_dlpack.to_dlpack(value)  # pyright: ignore
+    tensor = jax_dlpack.from_dlpack(tensor)
+    return tensor
+
+
+@torch_to_jax.register(abc.Mapping)
+def _torch_dict_to_jax(
+    value: Dict[str, Union[torch.Tensor, Any]]
+) -> Dict[str, Union[DeviceArray, Any]]:
+    """Converts a dictionary of PyTorch Tensors into a Dictionary of Jax DeviceArrays."""
+    return type(value)(**{k: torch_to_jax(v) for k, v in value.items()})
```

Comment:
This and the subsequent reversal function for each wrapper type is what is capturing the `Tuple`s. The typing may need to be updated to reflect that or an explicit `Tuple` function could be added?